### DISCUSSION TITLE
In extension methods should work for enums

### DISCUSSION THREAD

**Discussion by mburbea** [dotnet/csharplang#4562] [Language Ideas]
I was writing an extension method and I discovered something odd.

Consider the following example:
```cs
public static class C
{
        public enum E { }
        public static E M(this in E e) => e;
}
```
This shows the following error:
![image](https://user-images.githubusercontent.com/213233/111811624-a2fad780-88ad-11eb-8361-0dc662f165b1.png)

Enums are concrete value types, so I feel this error is incorrect

**Comment by bernd5** [comment]
Indeed, it seems to be a missing implementation for by in-ref extensions methods for enums. It should be moved to roslyn, or?
I found nothing about enum-extensions in the [readonly-ref](https://github.com/dotnet/csharplang/blob/1b4f6ac09fd0f29593cc8bf96244e2b16e2509cd/proposals/csharp-7.2/readonly-ref.md) proposal.
But under the hood they are just structs, too. Therefore it should work for them, too.

**Comment by 333fred** [comment]
FWIW, while I don't necessarily have a problem with allowing this, there is overhead in passing a value by ref instead of by value, and you're likely to be better-served by just passing the enum by value. `in` is generally best-used for larger structs that would be expensive to pass by value, while an enum is going to fit into a single register on most modern platforms.

**Reply by bernd5** [reply]
Yes, you are right. But it works for `byte`, too. Even if it makes no sense.

**Reply by 333fred** [reply]
Sure. As I said, I have no problem with allowing it, and am asking internally to see if this is a bug or was intentional :). But I wanted to point out that, besides language regularity, there is little point to enabling it as you're usually not going to want to use it.

**Reply by mburbea** [reply]
Also of note is that instance methods on struct are passed by-ref at the moment. E.g.
```cs
public static class C
{
        public readonly struct S
        {
            public S Foo() => this;
        }
        delegate S D(ref S a);
        public D Del {get;} = (D)Delegate.CreateDelegate(typeof(D), null, typeof(S).GetMethod(nameof(S.Foo)));
}
```

So I feel like if I wanted to make an extension method consistent with the calling conventions used elsewhere it should be at minimum ref.

**Reply by CyrusNajmabadi** [reply]
Fred is right here. This is just an implementation mistake. However, you really shouldn't do this. There is no upside, and a very real cost downside.

**Reply by BreyerW** [reply]
> 
> 
> Yes, you are right. But it works for `byte`, too. Even if it makes no sense.

actually passing single byte by ref has genuine use cases - eg that byte can be treated as memory address of array via `Unsafe` shenanigans

**Reply by CyrusNajmabadi** [reply]
I'm that case, you can just pass by ref.

**Reply by tannergooding** [reply]
There are use cases for passing by ref yes, but when the value is not actually being mutated then it is less efficient and can actually hurt perf (especially when the method isn't inlined) because the JIT has to treat it as address taken.

**Reply by tannergooding** [reply]
For cases like Hardware Intrinsics, we actually explicitly changed many of the methods off being instance and onto being by value extension methods instead to overcome the issue of instance methods being passed as `in` (or `ref` for non-readonly structs).

Just compare what we expose on:
https://source.dot.net/#System.Private.CoreLib/Vector128.cs,39cf292fbdb3dfaf vs https://source.dot.net/#System.Private.CoreLib/Vector128_1.cs,a7d647b6968ad0f9

You'll see that the only instance methods we have are ones where C# exposes a fundamental limitation on making it an extension method.

**Reply by mburbea** [reply]
That's good insight, thank you. I still think it should be allowed even if it isn't particularly beneficial in the usual case.

**Reply by theunrepentantgeek** [reply]
Keep in mind @mburbea that language changes come at a very high cost - they need formal specifications, implementations, tests (lots of tests!) and ongoing maintenance.

A feature needs to be _highly beneficial_ to be worth the effort - especially given the opportunity cost that other features can't be progressed while someone is working on this.

This is often referred to as [Every feature starts with -100 points](https://docs.microsoft.com/en-us/archive/blogs/ericgu/minus-100-points).

**Reply by mburbea** [reply]
@theunrepentantgeek, my understanding of this feature is that it's lack of implementation is an oversight. Not an intentional restriction. See both Fred's and Cyrus's comments earlier. Now I agree, it's not a particularly high priority. Outside of places where you want to mutate the ref, or do unsafe manipulations, its probably a performance downgrade unless you can be sure it'll inline.
